File: send_vite.py
#!/usr/bin/env python
# coding: utf-8
#



from accountBlock import AccountBlock, BlockType
import json
import os
from dotenv.main import load_dotenv
import requests
from common import Common

session = requests.Session()
session.trust_env = False

# Loads the .env file that resides on the same level as the script.
load_dotenv()

rpc_url = os.getenv('rpc_url')
rpc_timeout = float(os.getenv('rpc_timeout'))
token_id = os.getenv('token_id')

print (f"RPC_URL is {rpc_url}")

def json_rpc(rpc_url, payload):
    Common.logger.debug(f"Payload: {json.dumps(payload)}")
    response = session.post(rpc_url, json=payload, timeout=rpc_timeout).json()
    Common.logger.debug(f"Response: {json.dumps(response)}")
    return response

def get_previous_account_block(address):
    ab = {
        "jsonrpc": "2.0",
        "id": 17,
        "method": "ledger_getLatestAccountBlock",
        "params": [
            address
        ]
    }
    # Send request
    return json_rpc(rpc_url, ab)


def get_account_info(address):
    ab = {
        "jsonrpc": "2.0",
        "id": 17,
        "method": "ledger_getAccountInfoByAddress",
        "params": [
            address
        ]
    }
    # Send request
    return json_rpc(rpc_url, ab)

def get_account_balance(address):

    response = get_account_info(address)

    if "error" in response:
        raise Exception(f"Error grabbing acount info: {response}")

    result = response['result']  
    balanceInfo = result['balanceInfoMap']   
    viteInfo = balanceInfo['tti_5649544520544f4b454e6e40']
    balance = float(viteInfo['balance'])
    return Common.rawToVite(balance)


def get_account_quota(address):

    ab = {
        "jsonrpc": "2.0",
        "id": 1,
        "method": "contract_getQuotaByAccount",
        "params": [
            address,
        ]
    }
    # Send request
    response = json_rpc(rpc_url, ab)

    if "error" in response:
        raise Exception(f"Error grabbing quota info: {response}")

    result = response['result']  
    quota = int(result['currentQuota'])

    return Common.quotaToUT(quota)

# Return PoW difficulty for sending transaction
def get_pow_difficulty(address,toAddress,prevHash,data=""):
    
    ab = {
        "jsonrpc": "2.0",
        "id": 1,
        "method": "ledger_getPoWDifficulty",
        "params": [{
            "address": address,
            "previousHash": prevHash,
            "blockType": 2,
            "toAddress": toAddress,
            "data":data
        }]
    }
    # Send request
    return json_rpc(rpc_url, ab)


# _send_vite function with private key
def _send_vite(from_address, to_address, amount):

    balance = get_account_balance(from_address)
    print(f"Account: {from_address} Balance: {balance}")
    
    # Make sure that we have enough funds to cover transaction
    if(amount >= balance):
        raise Exception(f"Insufficient funds. Balance: {balance} Amount requested: {amount}")

    # Call send_vite.js script

File: send_vite.js
#!/usr/bin/env node
const vite = require("vitejs-notthomiz")
const HTTP_RPC = require("vitejs-notthomiz-http").default
const WS_RPC = require("vitejs-notthomiz-ws").default
const config = require("./config.json")
const BigNumber = require("bignumber.js").default

const [,,
    destination, amount
] = process.argv
if(!vite.wallet.isValidAddress(destination)){
    throw new Error("Invalid Address")
}
if(!/^\d+(\.\d+)?$/.test(amount)){
    throw new Error("Invalid Amount")
}

const url = new URL(config.VITE_NODE)
const provider = /^wss?:$/.test(url.protocol) ? 
    new WS_RPC(config.VITE_NODE, 6e5, {
        protocol: "",
        headers: "",
        clientConfig: "",
        retryTimes: Infinity,
        retryInterval: 10000
    }) : /^https?:$/.test(url.protocol) ? 
    new HTTP_RPC(config.VITE_NODE, 6e5) :
    new Error("Invalid node url: "+config.VITE_NODE)
if(provider instanceof Error)throw provider
console.log("Connecting to "+config.VITE_NODE)


const ViteAPI = new vite.ViteAPI(provider, async () => {
    console.log("Provider ready !")
    let address
    switch(config.VITE_LOGIN.type){
        case "mnemonic": {
            config.VITE_LOGIN.type = "seed"
            config.VITE_LOGIN.credentials = vite.wallet.getSeedFromMnemonics(config.VITE_LOGIN.credentials).seedHex
        }
        case "seed": {
            config.VITE_LOGIN.type = "private_key"
            config.VITE_LOGIN.credentials = vite.wallet.deriveKeyPairByIndex(config.VITE_LOGIN.credentials, config.VITE_LOGIN.index).privateKey
            config.VITE_LOGIN.index = 0
        }
        case "private_key": {
            if(config.VITE_LOGIN.index !== 0)throw new Error("Invalid index with private key: "+config.VITE_LOGIN.index)
            address = vite.wallet.createAddressByPrivateKey(config.VITE_LOGIN.credentials)
            break
        }
        default: {
            throw new Error("Invalid configuration for VITE_LOGIN")
        }
    }
    console.log("Using "+address.address+" as sender !")
    const actions = {
        send: async (tokenId, amount, destination) => {
            if(
                [tokenId, amount, destination].find(e => typeof e !== "string") ||
                !vite.utils.isValidTokenId(tokenId) ||
                !/^\d+$/.test(amount) ||
                !vite.wallet.isValidAddress(destination)
            )throw new Error("Invalid Arguments.")
            const balances = (await ViteAPI.request("ledger_getAccountInfoByAddress", address.address))?.balanceInfoMap || {}
            const balance = new BigNumber(balances[tokenId]?.balance || "0")
            if(balance.isLessThan(amount))throw new Error("Insufficient Balance")

            const accountBlock = vite.accountBlock.createAccountBlock("send", {
                toAddress: destination,
                address: address.address,
                tokenId: tokenId,
                amount: amount
            })
            accountBlock.setProvider(ViteAPI)
            .setPrivateKey(address.privateKey)
            const [
                quota,
                difficulty
            ] = await Promise.all([
                ViteAPI.request("contract_getQuotaByAccount", address.address),
                accountBlock.autoSetPreviousAccountBlock()
                .then(() => ViteAPI.request("ledger_getPoWDifficulty", {
                    address: accountBlock.address,
                    previousHash: accountBlock.previousHash,
                    blockType: accountBlock.blockType,
                    toAddress: accountBlock.toAddress,
                    data: accountBlock.data
                }))
            ])
            
            const availableQuota = new BigNumber(quota.currentQuota)
            if(availableQuota.isLessThan(difficulty.requiredQuota)){
                await accountBlock.PoW(difficulty.difficulty)
            }
            await accountBlock.sign()

            const hash = (await accountBlock.send()).hash

            return {
                hash: hash,
                from: address.address,
                to: destination,
                tokenid: tokenId,
                amount: amount
            }
        }
    }

    try{
        const result = await actions.send("tti_5649544520544f4b454e6e40", new BigNumber(amount).shiftedBy(18).toFixed().split(".")[0], destination)
        console.log(result)
        process.exit(0)
    }catch(err){
        console.error(err)
        process.exit(1)
    }
})